consensus/ibft/ibft.go

```diff
@@ -422,6 +423,10 @@ func (i *Ibft) buildBlock(snap *Snapshot, parent *types.Header) (*types.Block, e
 	}
 
 	_, root := transition.Commit()
+	// If we are the proposer, these events will be added in Validate state again
+	// No need to double add them here and have complicated logic in unstaking
+	staking.GetStakingHub().ClearEvents()
```

Comment:
Doesn't this mean that we might build a block with invalid (or exploiting) staking transactions, send `PREPREPARE` and `PREPARE` messages to other nodes, and only after that discover the invalid transactions in the `VALIDATE` state?

Can you elaborate a bit on this, how would that play out?

Reply:
As a proposer we go through the exact same transaction execution / block validation process as any other node, but we do it seemingly twice.

If the transaction is faulty / invalid, it should get handled accordingly and this is out of the scope of the staking hub.
We disregard all events regardless when initially building the block.